drivers/adc/adm1177/adm1177.c

```diff
@@ -0,0 +1,282 @@
+ *
+ * @param reg_address - In case of a write extended command byte operation
+ * 			one of the ADM1177_REG_ALERT_EN, ADM1177_REG_ALERT_TH
+ * 			or ADM1177_REG_CONTROL will be used.
+ * 			In case of a write command byte operation reg_address
+ * 			will be set to 0.
+ *
+ * @return 0 in case of success, -1 otherwise.
+******************************************************************************/
+int adm1177_write(struct adm1177_dev *device, uint8_t cmd, uint8_t reg_address)
+{
+	if(reg_address & ADM1177_EXT_CMD_MASK) {
+		uint8_t data[2] = {reg_address, cmd};
+
+		return no_os_i2c_write(device->i2c_desc, data, 2, 1);
+	} else {
+		if(cmd & (ADM1177_CMD_I_CONT | ADM1177_CMD_I_ONCE)) {
+			device->last_command = ADM1177_CURRENT_EN;
+			return no_os_i2c_write(device->i2c_desc, &cmd, 1, 1);
+		}
+		if(cmd & (ADM1177_CMD_V_CONT | ADM1177_CMD_V_ONCE)) {
+			device->last_command = ADM1177_VOLTAGE_EN;
+			return no_os_i2c_write(device->i2c_desc, &cmd, 1, 1);
+		}
+		if(cmd & (ADM1177_CMD_I_CONT | ADM1177_CMD_V_CONT) |
+		    cmd & (ADM1177_CMD_I_ONCE | ADM1177_CMD_V_ONCE)) {
+			device->last_command = ADM1177_VOLTAGE_AND_CURRENT_EN;
+			return no_os_i2c_write(device->i2c_desc, &cmd, 1, 1);
+		}
+		if(cmd & ADM1177_CMD_STATUS_RD) {
+			device->last_command = ADM1177_STATUS_EN;
+			return no_os_i2c_write(device->i2c_desc, &cmd, 1, 1);
+		}
+		if(cmd & ADM1177_CMD_VRANGE) {
+			device->vrange = ADM1177_VRANGE_HIGH;
+		} else {
+			device->vrange = ADM1177_VRANGE_LOW;
+		}
+		return -EINVAL;
+	}
+}
+
+/**************************************************************************//**
+ * @brief - Gets the voltage/current readback from the ADM1177s I2C ADC,
+ *	    if the device is initialized right and transmits the requested
+ *	    values.
+ *
+ * @param device - The device structure.
+ *
+ * @param conv_voltage - The variable through which the voltage read from the
+ * 			 ADC is received if requested.
+ * 			 If it is 0, it means it was not requested.
+ *
+ * @param conv_current - The variable through which the current read from the
+ * 			 ADC is received if requested.
+ * 			 If it is 0, it means it was not requested.
+ *
+ * @return 0 in case of success, -1 otherwise.
+******************************************************************************/
+int adm1177_readback(struct adm1177_dev *device,
+		     uint16_t* conv_voltage,
+		     uint16_t* conv_current)
+{
+	int ret;
+
+	uint8_t data[3];
+
+	switch(device->last_command) {
+	case ADM1177_CURRENT_EN:
+		ret = no_os_i2c_read(device->i2c_desc, data, 2, 1);
+		if(ret)
+			return ret;
+
+		/* First byte received contains the 8 MSBs of the
+		   current and is left shifted by 4, then the second
+		   byte contains the 4 LSBs of the current and is right
+		   shifted by 4, and when the OR operation between the
+		   two bytes happen we obtain the a 12-bit value of the
+		   current. */
+		*conv_current = (data[0] << 4) | (data[1] >> 4);
```

Comment:
if the user calls the function with adm1177_readback(dev, NULL, ptr) your program will crash because you are dereferencing NULL pointer

```
if (conv_current == NULL)
    return -EINVAL;

*conv_current = ...;
    
```

Reply:
Will change, and will add a specific function for read status.

Reply:
move this test before the i2c_read as well